llama-index-core/llama_index/core/agent/runner/base.py

```diff
@@ -702,3 +729,134 @@ async def astream_chat(
     def undo_step(self, task_id: str) -> None:
         """Undo previous step."""
         raise NotImplementedError("undo_step not implemented")
+
+
+class BasePlanningAgentRunner(AgentRunner):
+    @abstractmethod
+    def create_plan(self, input: str, **kwargs: Any) -> str:
+        """Create plan. Returns the plan_id."""
+        ...
+
+    @abstractmethod
+    def get_next_tasks(self, plan_id: str, **kwargs: Any) -> List[str]:
+        """Get next task ids for a given plan."""
+        ...
+
+    @abstractmethod
+    def mark_task_complete(self, plan_id: str, task_id: str, **kwargs: Any) -> None:
+        """Mark task complete for a given plan."""
+        ...
+
+    @abstractmethod
+    def refine_plan(self, input: str, plan_id: str, **kwargs: Any) -> None:
+        """Refine plan."""
+        ...
+
+    @abstractmethod
+    def run_task(self, task_id: str, **kwargs: Any) -> TaskStepOutput:
+        """Run task."""
+        ...
+
+    async def acreate_plan(self, input: str, **kwargs: Any) -> str:
+        """Create plan (async). Returns the plan_id."""
+        return self.create_plan(input, **kwargs)
+
+    async def arefine_plan(self, input: str, plan_id: str, **kwargs: Any) -> None:
+        """Refine plan (async)."""
+        return self.refine_plan(input, plan_id, **kwargs)
+
+    async def arun_task(self, task_id: str, **kwargs: Any) -> TaskStepOutput:
+        """Run task (async)."""
+        return self.run_task(task_id, **kwargs)
+
+    @dispatcher.span
+    def _chat(
+        self,
+        message: str,
+        chat_history: Optional[List[ChatMessage]] = None,
+        tool_choice: Union[str, dict] = "auto",
+        mode: ChatResponseMode = ChatResponseMode.WAIT,
+    ) -> AGENT_CHAT_RESPONSE_TYPE:
+        """Chat with step executor."""
+        dispatch_event = dispatcher.get_dispatch_event()
+
+        if chat_history is not None:
+            self.memory.set(chat_history)
+
+        # create initial set of tasks
+        plan_id = self.create_plan(message)
+
+        results = []
+        dispatch_event(AgentChatWithStepStartEvent())
+        while True:
+            # EXIT CONDITION: check if all sub-tasks are completed
+            next_task_ids = self.get_next_tasks(plan_id)
+            if len(next_task_ids) == 0:
+                break
+
+            jobs = [
+                self.arun_task(sub_task_id, mode=mode, tool_choice=tool_choice)
+                for sub_task_id in next_task_ids
+            ]
+            results = asyncio.run(asyncio.gather(*jobs))
```

Comment:
is it possible to use async_utils.run_jobs for this? Doing so will correctly tie events/spans when since there is some instrumentation logic baked into that util func.

Reply:
ah yea I can, I just didn't want to limit the number of jobs or make it configurable lol

Reply:
lol I hear you